File: src/net/callumtaylor/asynchttp/processor/StringProcessor.java
package net.callumtaylor.asynchttp.processor;

public class StringProcessor extends Processor<String>
{
	private StringBuffer stringBuffer;

	@Override public void onPublishedDownloadProgress(byte[] chunk, int chunkLength, long totalProcessed, long totalLength)
	{
		if (stringBuffer == null)
		{
			int total = (int)(totalLength > Integer.MAX_VALUE ? Integer.MAX_VALUE : totalLength);
			stringBuffer = new StringBuffer(Math.max(8192, total));
		}

		if (chunk != null)
		{
			try
			{
				stringBuffer.append(new String(chunk, 0, chunkLength, "UTF-8"));
			}
			catch (Exception e)
			{
				e.printStackTrace();
			}
		}
	}

	/**
	 * Processes the response from the stream.
	 * This is <b>not</b> ran on the UI thread
	 *
	 * @return The data represented as a String
	 */
	@Override public String getContent()
	{
		return stringBuffer.toString();
	}
}

File: kotlin/src/main/java/net/callumtaylor/asynchttp/processor/StringProcessor.kt
package net.callumtaylor.asynchttp.processor

import net.callumtaylor.asynchttp.AsyncHttpClient
import java.io.InputStream

/**
 * // TODO: Add class description
 */
open class StringProcessor : ResponseProcessor<String>()
{
	override fun processStream(inputStream: InputStream, contentLength: Long, progressCallback: (Long, Long) -> Unit): String
	{
		val buffer = ByteArray(AsyncHttpClient.BUFFER_SIZE)
		val string = StringBuffer()

		var len = 0
		var readCount = 0L
		try
		{
			inputStream.use { stream ->
				len = stream.read(buffer)

				if (len > -1)
				{
					progressCallback(readCount, contentLength)
					readCount += len

					string.append(String(buffer, 0, len))
				}
			}

			inputStream.close()
		}
		catch (e: Exception){}

		progressCallback(readCount, readCount)

		return string.toString()
	}
}